### ISSUE TITLE
patchGlobal() callback does not fire when client is disabled

### ISSUE THREAD

**Issue by mikeybtn** [getsentry/raven-node#138]
In the following example, the callback will never fire:

``` js
var dsn = false;  // disabled
var client = new raven.Client(dsn);

client.patchGlobal(function() {
  console.log('Uncaught exception, aborting!');
  process.exit(1);
});
```

We were treating the callback passed to `client.patchGlobal` as our process's default uncaught exception handler, expecting it to be invoked regardless of whether the sentry client is reporting, so this behavior was a little surprising.

Potential improvements:
- Always fire the callback, adding a `result` parameter which indicates whether anything was logged.
- Just document the behavior: If the client is disabled, the callback will never fire.

I'm happy to prepare a patch if there's a recommendation. Thanks!

**Comment by benvinegar** [comment]
Thanks for pointing this out.

> Just document the behavior: If the client is disabled, the callback will never fire.

I think this is our preference. I wouldn't advise people to depend on Raven's `patchGlobal` callback as a global error handler.

**Comment by LoicMahieu** [comment]
The example could not works:
```js
process.on('uncaughtException', function (err) {
  console.error(err)
  process.exit(1)
})
client.patchGlobal()
```
As the process will be exited before Raven have the time to complete send the request to Sentry.
So if the callback of `patchGlobal` is not fired when Raven is disabled, @benvinegar how exit the process if we couldn't depend on Raven's `pathGlobal` ?

Current hack is to not `pathGlobal` when no dsn:
```js
if (dsn) {
  client.patchGlobal(function() {
    console.log('Uncaught exception, aborting!');
    process.exit(1);
  });
}
```

**Comment by DamonGant** [comment]
@benvinegar This is not entirely clear from the documentation, in fact I'm getting the opposite impression from the docs.

    Passing any falsey value as the DSN will disable sending events _upstream_

https://docs.getsentry.com/hosted/clients/node/usage/#disable-raven

@LoicMahieu You can capture the errors manually and exit the process once you get the first event from the client. Check https://github.com/getsentry/raven-node/blob/master/lib/client.js#L238

**Comment by migueloller** [comment]
Now that `patchGlobal` is `install`, what should we do?

It would make sense that the callback is called even if the event isn't sent to Sentry because of what @LoicMahieu mentioned [here](https://github.com/getsentry/raven-node/issues/138#issuecomment-205282008).

Also, the arguments passed to the callback should be documented. It looks like the error is being passed somewhere in there, which could be used for logging and avoid the error from happening silently.

**Comment by migueloller** [comment]
Just noticed that it looks like that callback is called for unhandled rejections (promises) even if there's no DSN. It would make more sense for the uncaught exceptions behavior to be consistent with this.

**Comment by schittler** [comment]
I think the functionality itself encourages bad code. I'd favour deprecating `patch global` (or whatever it is called now) and update the documentation sample code with with an actual global event handler that can do things beyond capturing events for Sentry. It is simply out of scope and right now I'm getting the impression I can just wire up raven and never worry about error handling again.

The code should `especially` mention, that ignoring a throw in node, especially one caused by native code can have [unexpected side effects and the process is often better off restarted](https://shapeshed.com/uncaught-exceptions-in-node/#let-your-application-crash-log-and-restart).

**Comment by LewisJEllis** [comment]
> Just noticed that it looks like the callback is called for unhandled rejections (promises) even if there's no DSN. So it would make sense for the uncaught exceptions behavior to be consistent with this.

Good point, that was probably an oversight. I'll make sure this ends up consistent one way or the other.

> an actual global event handler that can do things beyond capturing events for Sentry.

The difficulty is that the global handler that captures events for sentry isn't as simple as
```javascript
process.on('uncaughtException, function (err) {
  Raven.captureException(err, function () {
    process.exit(1);
  });
});
```
because we have to be careful to avoid some recursive error cases (and `process.on('uncaughtException')` isn't the only place we see errors surface). I definitely don't want to put the expectation on the user to roll their own global handler for reporting to Sentry - completely defeats the ease of installation we're going for - but they should certainly be able to know when raven detects/captures an exception.

Right now that situation isn't completely coherent, as outlined in #257; this also ties into discussion there on what the role of the `install(cb)` callback should be. Should that callback be treated as a global error handler, where we can expect raven to invoke it in all fatal exception situations such that users may rely on it and use it to clean up/exit? Should that be the recommended/required way to go about setting up a global error handler when using Raven? Or should Raven be more hands-off here and stay behind the scenes? If so, what should that look like?

In general, I think I'm in favor of letting this `install` callback play a bigger role in the overall error handling scheme and allow users to rely on it more as seems to be desired here, but it could be nice to avoid locking people in to using/relying on it, and I definitely don't want to require users to write their own `install` callback (or other global error handler) for typical basic program operation.

**Comment by migueloller** [comment]
@LewisJEllis, I agree with the sentiment that if you're going for an easy setup it would make sense to have the `install` callback play that bigger role. And it looks like it already kind of does:

![image](https://cloud.githubusercontent.com/assets/5677929/22114906/0698d130-de39-11e6-8179-f6adad97f33b.png)

Being able to simply remove lines 16-22 would be awesome. Especially when the `install` callback already does it for `process.on('unhandledRejection')`.

**Comment by LewisJEllis** [comment]
In #308 we went the direction of what was previously mentioned here:
> letting this install callback play a bigger role in the overall error handling scheme and allow users to rely on it more

So the current behavior is that this:
```javascript
var Raven = require('raven');
Raven.config(false).install(function (err) {
  console.log('fatal error callback');
});
throw new Error('derp');
```
outputs:
```
raven@1.2.1 alert: no DSN provided, error reporting disabled
fatal error callback
```
and this:
```javascript
var Raven = require('raven');
Raven.config(false).install();
throw new Error('derp');
```
outputs:
```
raven@1.2.1 alert: no DSN provided, error reporting disabled
Error: derp
    at Object.<anonymous> (...)
    <rest of stack trace>
```
since passing no callback to `install()` will default to behave as if you passed:
```javascript
Raven.config(...).install(function (err, sendErr, eventId) {
  console.error(err && err.stack ? err.stack : err);
  process.exit(1);
});
```

As for the other concern mentioned here:
> As the process will be exited before Raven have the time to complete send the request to Sentry

The primary case where this was actually an issue was also addressed in #308 (some more details in #127). So I consider this issue resolved.
